{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Goal of BioCode: To create a Phylogenic Tree using data from GenBank and use the excel sheet as the overlay\n",
    "\n",
    "import csv\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import sys\n",
    "import os\n",
    "import ssdeep as ss\n",
    "import hashlib,binascii\n",
    "import tensorflow as tf\n",
    "import xlsxwriter\n",
    "from sklearn import preprocessing\n",
    "from keras.layers import Input, Dense\n",
    "from keras.models import Model, Sequential\n",
    "from Bio import SeqIO\n",
    "from Bio import Phylo\n",
    "from scipy.cluster.hierarchy import dendrogram, linkage\n",
    "from matplotlib import pyplot as plt\n",
    "\n",
    "class organism:\n",
    "    \n",
    "    def __init__(self, name, hashVal):\n",
    "        self.name = name\n",
    "        self.hashVal = hashVal\n",
    "\n",
    "# This class variable is for the list later to contain the 4 letter sequences (Not necessary here)\n",
    "#class nucleotideSequence:\n",
    "    \n",
    "#    def __init__(self, seq):\n",
    "#        self.seq = seq\n",
    "    \n",
    "\n",
    "# This function will read the columns from the excel file and delete NaN value columns (Not necessary here)\n",
    "# input: file=name of the file, dropColumns=Which columns to drop (default is none), debug=print values (default is 0)\n",
    "# return: df=dataframe created from the excel file\n",
    "def convertExcelToDataFrame(file, dropColumns=None, debug=0):\n",
    "\n",
    "    raw_data = pd.read_excel(file)\n",
    "    df = pd.DataFrame(raw_data)\n",
    "    df = df.dropna(axis=\"columns\", how=\"all\")\n",
    "\n",
    "    if dropColumns != None:\n",
    "        df = df.drop(dropColumns, axis=1)\n",
    "        \n",
    "    df = df.dropna(axis=0)\n",
    "    \n",
    "    if debug == 1:\n",
    "        print(df)\n",
    "\n",
    "    return df\n",
    "\n",
    "# This function will take a dataframe and convert the string data into numeric data (Not necessary here)\n",
    "# Unused function\n",
    "def oneHotEncodeData():\n",
    "\n",
    "    df_categorical = df[[\"Strain\", \"Genome size\"]]\n",
    "                \n",
    "    df_numeric = df.select_dtypes(exclude=[object])\n",
    "\n",
    "    encode = preprocessing.OneHotEncoder(handle_unknown=\"ignore\")\n",
    "    encode.fit(df_categorical)\n",
    "    one_hot_labels = encode.transform(df_categorical).toarray()\n",
    "\n",
    "    all_data = np.concatenate((one_hot_labels, df_numeric.values), axis=1)    \n",
    "    \n",
    "# This function will read the fasta files given and returns a hash value for each file\n",
    "# input: files=array of files given, debug=print values (default is 0)\n",
    "def readFastaFiles(files, debug=0):\n",
    "    \n",
    "    if debug == 1:\n",
    "        print(files)\n",
    "        print(\"\")\n",
    "        \n",
    "    organisms = []\n",
    "    sequenceString = \"\"\n",
    "    \n",
    "    if type(files) != list:\n",
    "        print(\"Error, Expected a list but was given %s\" % (type(files) ) )\n",
    "        sys.exit()\n",
    "\n",
    "    try:\n",
    "        for file in files:\n",
    "            for record in SeqIO.parse(file, \"fasta\"):\n",
    "                name = \"\"\n",
    "                sequenceString = \"\"\n",
    "                for letter in record.seq:\n",
    "                    sequenceString+=letter\n",
    "                for letter in record.id:\n",
    "                    name+=letter\n",
    "                hashVal = ss.hash(sequenceString)\n",
    "                organisms.append(organism(name, hashVal) );\n",
    "    \n",
    "    except ValueError as e:\n",
    "        print(\"Value error ({0}): {1}\".format(e.errno, e.strerror))\n",
    "    except IOError as e:\n",
    "        print(\"I/O error ({0}): {1}\".format(e.errno, e.strerror))\n",
    "    \n",
    "    return organisms\n",
    "\n",
    "\n",
    "# This function will create a list out of all files within the directory given\n",
    "def retrieveFastaFiles(folder):\n",
    "    fileList = []\n",
    "    \n",
    "    for file in os.listdir(folder):\n",
    "        fileList.append(folder + file)\n",
    "    \n",
    "    return fileList\n",
    "\n",
    "# Create an XML file using an organism's name and hash value of genetic data\n",
    "def createXMLFile(organisms, fileName):\n",
    "    \n",
    "    row = 0\n",
    "    col = 0\n",
    "    \n",
    "    workbook = xlsxwriter.Workbook(fileName)\n",
    "    worksheet = workbook.add_worksheet()\n",
    "    \n",
    "    for organismm in organisms:\n",
    "        worksheet.write(row, col, organismm.name)\n",
    "        worksheet.write(row, col+1, organismm.hashVal)\n",
    "        row+=1\n",
    "        \n",
    "    workbook.close()\n",
    "\n",
    "# Task for building the phylogenetic tree:\n",
    "# 1) Create a matrix using comparisons of the different hashes DONE\n",
    "# 2) Numbers will be 0-100, determine which organisms will be connected based off similarity\n",
    "# 3) Use Scipy from the information gathered to create the tree\n",
    "    \n",
    "def identityMatrixOfHashes(dataFrame, debug=0):\n",
    "    \n",
    "    i = 0\n",
    "    j = dataFrame.shape[0]\n",
    "    arrayOfHashes = []\n",
    "    matrix = [ [ 0 for i in range(dataFrame.shape[0]) ] for j in range(dataFrame.shape[0]) ]\n",
    "    \n",
    "    for i in range(0, dataFrame.shape[0]):\n",
    "        #print(dataFrame.iloc[i]['Hash Values'])\n",
    "        arrayOfHashes.append(dataFrame.iloc[i]['Hash Values'])\n",
    "    \n",
    "    \n",
    "    \n",
    "    if debug == 1:\n",
    "        print(arrayOfHashes)\n",
    "    \n",
    "    for i in range(0, dataFrame.shape[0]):\n",
    "        for j in range(0, dataFrame.shape[0]):\n",
    "            matrix[i][j] = ss.compare(arrayOfHashes[i], arrayOfHashes[j])\n",
    "            \n",
    "    if debug == 1:\n",
    "        for array in matrix:\n",
    "            print(array)\n",
    "            \n",
    "    return matrix\n",
    "        \n",
    "def scipyTree(identityMatrix):\n",
    "    \n",
    "    Z = linkage(identityMatrix, 'ward')\n",
    "    fig = plt.figure(figsize=(25, 10) )\n",
    "    dn = dendrogram(Z)\n",
    "    \n",
    "    Z = linkage(identityMatrix, 'single')\n",
    "    fig = plt.figure(figsize=(25, 10) )\n",
    "    dn = dendrogram(Z)\n",
    "    plt.show()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "fileList = retrieveFastaFiles(\"./allFixedFastaFiles/\")\n",
    "\n",
    "organisms = readFastaFiles(fileList)\n",
    "\n",
    "createXMLFile(organisms)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
